### ISSUE TITLE
Custom task fails to load in Windows.

### ISSUE THREAD

**Issue by JunielKatarn** [Microsoft/msbuild#1483]
Commit: 39c48801426a0aa7a18dbb43a981dd6af8765881
OS: Windows 10 (1607)

Note: This issue does not show up in UNIX (Ubuntu, Mac OS).

I created a custom task using .NET core (1.0.0-preview2-003121): `MyTask`.
Here is the `project.json` for my custom task:
```
{
	"name": "MyTasks",
	"version": "1.0.0-*",
	"buildOptions": {
		"debugType": "portable",
		"emitEntryPoint": false
	},
	"dependencies": {
		//NOTE: Use these versions to avoid manifest mismatches.
		"Microsoft.Build.Framework": "0.1.0-preview-00024-160610",
		"Microsoft.Build.Utilities.Core": "0.1.0-preview-00024-160610",
		"System.Diagnostics.Process": "4.1.0"
	},
	"frameworks": {
		"netcoreapp1.0": {
			"dependencies": {
				"Microsoft.NETCore.App": {
					"type": "platform",
					"version": "1.0.0"
				}
			},
			"imports": "dnxcore50"
		}
	}
}
```

When running the msbuild instance built from source against a custom project, invoking my custom task, I get this error (trimmed):
```
error MSB4061: The "MyTask" task could not be instantiated from "C:\Build\...\MyTask.dll". \r [C:\Build\...\Source\Samples\application\application.proj]
error MSB4061: System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.\r [C:\Build\...\Source\Samples\application\application.proj]
error MSB4061: File name: 'System.Runtime, Version=4.1.0.0, Culture=neutral, PublicKey
Token=b03f5f7f11d50a3a'\r [C:\Build\...\Source\Samples\application\application.proj]
error MSB4061:    at MyTask..ctor()\r [C:\Build\...\Source\Samples\application\application.proj]
...
error MSB4060: The "MyTask" task has been declared or used incorrectly, or failed during construction. Check the spelling of the task name and the assembly name. [C:\Build\...\Source\Samples\application\application.proj]
```

This is how I imported the task in my MSBuild .proj file:
`<UsingTask TaskName="MyTask" AssemblyFile="$(MSBuildTempTasksPath)MyTask.dll" />`

**Comment by jeffkl** [comment]
I created a .NET core class library to attempt to repro this but it works fine.  

https://github.com/jeffkl/MSBuild-NetCore/tree/master/src/MyTask

Can you try changing your project.json to be more like a class library?  Target `netstandard1.3` and depend on `NETStandard.Library` 1.6.  

My example worked for me on Windows and Linux.  Let me know if you get it to work.

**Comment by JunielKatarn** [comment]
Hi! As far as getting around the first issue, it works fine. Thanks!
I'm curious, though. Why target `netstandard1.3` but depend on `NETStandard.Library 1.6`? Are those different ".NET Standard Libraries"?

Unfortunately, now I run into a different error.
My task code uses namespace `System.Diagnostics.Process`. Using the version matrix you suggest, forces the built DLLs to use version `4.1.0`. When using an msbuild instance (from this repository), I get a 'FileNotFound' error for that assembly (sys.diag.process), version `4.0.0`.
In short, those versions force building 4.1.0, but the current MSBuild sources need 4.0.0. I'm stuck.

**Comment by jeffkl** [comment]
> I'm curious, though. Why target netstandard1.3 but depend on NETStandard.Library 1.6? Are those different ".NET Standard Libraries"?

I know its confusing and it took me a while to understand it.  `netstandard1.3` is a set of reference libraries with a defined set of available APIs.  When you "target" that as a framework, you're only allowing your application to use APIs that are available for a runtime that has them.  The two main runtimes at the moment are the .NET Framework and .NET Core.  If you target `netstandard1.3`, your application can run in .NET Framework 4.6 or .NET Core 1.0.

The real confusing part is the NuGet package `NETStandard.Library 1.6`.  This NuGet package's version is completely orthogonal to the .NET you target.  If you look at the [NuGet page for the package](https://www.nuget.org/packages/NETStandard.Library) you'll see it contains reference assemblies for `netstandard1.0`, `netstandard1.1`, `netstandard1.2`, and `netstandard1.3`.  So the `framework` listed in your project.json is what tells NuGet what reference assemblies to use.  The fact that version of the package is `1.6` is extremely confusing (and I didn't do it :))

You only want to target `netcoreapp1.0` if you're making an executable and class libraries should always target a `netstandard` platform.

> In short, those versions force building 4.1.0, but the current MSBuild asks for 4.0.0.

Something else is going on then because the package [Microsoft.Build.Framework](https://www.nuget.org/packages/Microsoft.Build.Framework) does not depend on `System.Diagnostics.Process` and [Microsoft.Build.Utilities.Core](https://www.nuget.org/packages/Microsoft.Build.Utilities.Core/) depends on `System.Diagnostics.Process 4.1.0`.  I updated my sample code to use the same NuGet packages as your project.json and my task ran successfully on Windows and Linux.

https://github.com/jeffkl/MSBuild-NetCore/commit/cb93d8cdc7834468b793e78c016a4895b1229c56

I doubt this is related to anything MSBuild is doing so I should probably refer you to the [.NET Standard folks](https://github.com/dotnet/standard).

**Comment by JunielKatarn** [comment]
It was related to MSBuild. Turns out I was building the wrong favor (the default one, by the way).
Long story short:
MSBuild/UNIX always targets .NET Core (netstandard1.3, in this specific case).
MSBuild/Windows targets win7-x64 by default (See cibuild.cmd). Meaning, it targets a 'Full' .NET Framework, which is available only on Windows, and does consume some specific library versions not available/compatible with .NET Core.

The solution was building MSBuild like this:
`cibuild.cmd --target CoreCLR`
Then, instead of using a generated MSBuild.exe, use 'corerun.exe MSBuild.dll'.

I'll close this issue now, but I think this should be pointed out in the documentation. Took me quite a while to figure out.

Thanks for your help, @jeffkl.

@cdmihai, @rainersigwald, do you think this should be pointed in the documentation? (Windows builds default to full .NET Framework by default, risking incompatibilities with .NET Core libraries).
